### DISCUSSION TITLE
NeXus Data - myHDF5 - Expected some value

### DISCUSSION THREAD

**Discussion by joaonizer** [silx-kit/h5web#1668] [Q&A] [answered]
Hello,

I am saving some simulation results that include scattering parameters (complex data, real and imaginary). I'm creating a Group called SParameters with attribute NX_class:NX_entry, then subgroups for each S-Parameter (S11, .... , S22) with NX_class:NXdata, folowing an example file provided by [myHDF5](https://myhdf5.hdfgroup.org/), this onefor example: [Zenodo Example](https://zenodo.org/record/6497438/files/xrr_dataset.h5) or clicking in the Zenodo example here: [Zenodo Example myHDF5](https://myhdf5.hdfgroup.org/help). You can check my file down here, and use the inspect option to check the groups attributes and datasets:

**My dataset:**
[wg_straight_2024-06-11 15-20-21-960340_75feb3d1-8ca5-4be4-829d-01a50d8abd47.zip](https://github.com/user-attachments/files/15791388/wg_straight_2024-06-11.15-20-21-960340_75feb3d1-8ca5-4be4-829d-01a50d8abd47.zip)

**Inspect option:**
![image](https://github.com/silx-kit/h5web/assets/26409849/39e7332f-2077-4c93-ae2f-8392741139bf)

I am doing that to use the myHDF5 feature of plotting automatically the data, without navigating at each SNN dataset. Unfortuantely, whenever I click in a particular SNN:group of NX_class:NXdata, it does not automatically show the plot, and does not include axis names and all. It only show a mesage Expected some value, as you can see in the figure down here:

![image](https://github.com/silx-kit/h5web/assets/26409849/16b9becc-6466-4529-aa41-a3d69ac3c9eb)

I am using python's h5py module to create the HDF5 file and create/set groups, attributes and datasets. 

Any help is appreciated!

**Comment by axelboc** [comment]
Hi @joaonizer, I think it might be because of the syntax of the [`axes` attribute](https://manual.nexusformat.org/classes/base_classes/NXdata.html#index-12), which should contain an array of strings instead of a scalar string. Sorry for the meaningless error message... I'll try to debug where it comes from.

You can also check out the [mock demo](https://h5web.panosc.eu/mock) for examples on how to structure your NeXus file - e.g. `/nexus_entry/spectrum` (note that it's a 2D dataset visualized as _Line_, which is why the `axes` array has the extra `'.'`.)

Minor suggestion: you could also add a `default` attribute to the root group of your file to make sure that the right `NXdata`  group is rendered by default. This `default` attribute can simply point to `'/SParameters'`; H5Web will then know to follow `SParameters`' own `default` attribute. It works without it because `SParameters` is first, but it's always good to be more explicit.

**Reply by joaonizer** [reply]
I have added the "default" to the root. It now points to the SParameters entry. I noticed that the problem is related to the "axes" "Value".
I also added the [".", "/SParameters/wavelength/"]:

<image src="https://github.com/silx-kit/h5web/assets/26409849/eedaab55-7709-4d5f-a7ad-a0701dc813a7" height="300">

But the problem still persists. The same message is shown: Expected some Value.

I noticed though, that the DataSet used in the mock demo does contain a Value in its attribute:

<image src="https://github.com/silx-kit/h5web/assets/26409849/fc0d575d-eb37-478a-bb2d-6c203f440830" height="300">

Mine, however, does not contain that attribute showing the values:

<image src="https://github.com/silx-kit/h5web/assets/26409849/726e1061-3311-4e69-9c8f-3ba055264cc7" height="300">


Also, if I just remove the axes attribute, the plot happens automatically as it should, but then the Xaxis are just the index of each point.

Here is the data set with the added strings:
[with_axes.zip](https://github.com/user-attachments/files/15804811/with_axes.zip)

Here is the dataset without the axes argument:
[without_axes_plotting_auto.zip](https://github.com/user-attachments/files/15804815/without_axes_plotting_auto.zip)

Thanks for the updates so far! :)

Perhaps I'm still doing something silly, but I'm creating the datasets all the same way:
```python
with h5py.File(self.sim_name + ".hdf5", "w") as f:
    self.hdf5_file = f
    f.attrs["date_utc"] = date_utc
    f.attrs["sim_id"] = sim_id
    f.attrs["dx"] = dx
    f.attrs["dy"] = dy
    f.attrs["dz"] =dz
    f.attrs["dt"] = dt
    f.attrs["NX_class"] = "NXroot"
    f.attrs["default"] = "SParameters"
```

```python
with h5py.File(simulation.sim_name+".hdf5", "a") as f:
    grp = f.create_group("SParameters")
    grp.attrs["NX_class"] = "NXentry"
    grp.attrs["default"] = "S11"
    dataset = grp.create_dataset("frequency", data=f_range/1e12)
    dataset.attrs["long_name"] = ["Frequency"]
    dataset.attrs["units"] = ["THz"]
    dataset = grp.create_dataset("wavelength", data=l_range/1e-6)
    dataset.attrs["long_name"] = ["Wavelength"]
    dataset.attrs["units"] = ["um"]
```

```python
with h5py.File(simulation.sim_name+".hdf5", "a") as f:
    sp_hdf5 = mag_plot*np.exp(-1j*phase_plot)
    grp = f.create_group(f"SParameters/S11")
    grp.attrs["signal"] = "data"
    #grp.attrs["axes"] = [".","/SParameters/wavelength"]
    grp.attrs["NX_class"] = "NXdata"
    grp.attrs["interpretation"] = "spectrum"
    grp.attrs["SILX_style"] = "{\"axes_scale_types\": [\"linear\", \"linear\"], \"signal_scale_type\": \"linear\"}"
    grp.attrs["title"] = f"S11"
    grp.create_dataset("data", data=sp_hdf5)
```

**Reply by axelboc** [reply, marked as answer]
Sorry I wasn't clear. Since your dataset is 1D, the axes attribute should just have: `["wavelength"]`. **EDIT:** it should also just be the name of the child dataset to use as axis, not its absolute path.

Please disregard the `value` property in the mock demo. I will see if I can prevent it from being displayed.

**Reply by joaonizer** [reply]
Right, indeed if I create the dataset `wavelength` inside each SNN `NXData`, then the `axes` works fine with `["wavelength"]`. Is it possible to reuse the dataset I have above in `SParameters` for all the axes, instead of adding multiple datasets representing the same values for the axes?

**Reply by axelboc** [reply]
Glad to hear! Yes, for sure, you can create multiple [soft links](https://docs.h5py.org/en/stable/high/group.html#soft-links) to the same dataset.

**Reply by joaonizer** [reply]
Would you mind providing me with an example?

I've tried passing the ```h5py.SoftLink("SParameters/wavelength")``` instead of `"wavelength"` in `["wavelength"]`, but it did not work. h5py documentation does not expand on SoftLinks for Datasets.

**Reply by axelboc** [reply]
The `axes` attribute stays the same: `["wavelength"]`.

```python
grp["wavelength"] = h5py.SoftLink("/SParameters/wavelength")
```
